core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java

```diff
@@ -149,24 +164,132 @@ public void close() throws IOException {
+    if (mJournalOutputStream != null) {
+      return;
+    }
+    UfsJournalSnapshot snapshot = UfsJournalSnapshot.getSnapshot(mJournal);
+    UfsJournalFile currentLog = snapshot.getCurrentLog(mJournal);
+    if (currentLog == null) {
+      return;
+    }
+    long startSeq = currentLog.getStart();
+    long lastPersistSeq = -1;
+    LOG.info("Trying to recover from previous UFS journal failure."
+        + " Scanning for the first corrupted journal entry.");
+    try (JournalReader reader = new UfsJournalReader(mJournal, startSeq, true)) {
+      JournalEntry entry;
+      while ((entry = reader.read()) != null) {
+        if (entry.getSequenceNumber() > lastPersistSeq) {
+          lastPersistSeq = entry.getSequenceNumber();
+        }
+      }
+    } catch (InvalidJournalEntryException e) {
+      LOG.info("Found first corrupted journal entry.");
+    } catch (IOException e) {
+      LOG.info("I/O error encountered while trying to read journal.");
+      throw e;
+    }
+    String src = currentLog.getLocation().toString();
+    LOG.info("Last valid (persisted) journal entry sequence number in {} is {}",
+        src, lastPersistSeq);
+    if (lastPersistSeq < 0) {
+      mUfs.deleteFile(src);
+      return;
+    }
+
+    String dst = UfsJournalFile
+        .encodeLogFileLocation(mJournal, currentLog.getStart(), lastPersistSeq + 1).toString();
+    if (mUfs.exists(dst)) {
+      LOG.warn("Deleting duplicate completed log {}.", dst);
+      // The dst can exist because of a master failure during commit. This can only happen
+      // when the primary master starts. We can delete either the src or dst. We delete dst and
+      // do rename again.
+      mUfs.deleteFile(dst);
+    }
+    LOG.info("Renaming the previous incomplete journal file.");
+    mUfs.renameFile(src, dst);
+
+    maybeRotateLog();
+    if (!mEntriesToFlush.isEmpty()) {
+      JournalEntry entry = mEntriesToFlush.peek();
+      if (entry.getSequenceNumber() > lastPersistSeq + 1) {
+        throw new IOException(ExceptionMessage.JOURNAL_ENTRY_MISSING
+            .getMessageWithUrl(RuntimeConstants.ALLUXIO_DEBUG_DOCS_URL,
+            lastPersistSeq + 1, entry.getSequenceNumber()));
+      }
+      LOG.info("Retry writing unwritten journal entries");
+    }
+    for (JournalEntry entry : mEntriesToFlush) {
+      if (entry.getSequenceNumber() > lastPersistSeq) {
+        try {
+          entry.toBuilder().build()
+              .writeDelimitedTo(mJournalOutputStream.mOutputStream);
+        } catch (IOException e) {
+          mRotateLogForNextWrite = true;
+          UfsJournalFile tempLog = mJournalOutputStream.mCurrentLog;
+          mJournalOutputStream = null;
+          throw new IOException(ExceptionMessage.JOURNAL_WRITE_FAILURE
+          .getMessageWithUrl(RuntimeConstants.ALLUXIO_DEBUG_DOCS_URL,
+              tempLog, e.getMessage()), e);
+        }
+      }
+    }
+    if (!mEntriesToFlush.isEmpty()) {
```

Comment:
we can keep everything inside the earlier `!mEntriesToFlush.isEmpty()` block. Or to avoid indentation, return early if `mEntriesToFlush.isEmpty()` is true

Reply:
Done.